packages/appium/lib/schema/arg-spec.js

```diff
@@ -0,0 +1,223 @@
+ * We use this in the CLI to convert it to `argparse` options.
+ */
+export const APPIUM_CONFIG_SCHEMA_ID = 'appium.json';
+
+/**
+ * The schema prop containing server-related options. Everything in here
+ * is "native" to Appium.
+ * Used by {@link flattenSchema} for transforming the schema into CLI args.
+ */
+export const SERVER_PROP_NAME = 'server';
+
+/**
+ * Used to parse extension info from a schema ID.
+ */
+const SCHEMA_ID_REGEXP = /^(?<extType>.+?)-(?<normalizedExtName>.+)\.json$/;
+
+/**
+ * Avoid typos by using constants!
+ */
+const PROPERTIES = 'properties';
+
+/**
+ * An `ArgSpec` is a class representing metadata about an argument (or config
+ * option) used for cross-referencing.
+ *
+ * This class has no instance methods, and is basically just a read-only "struct".
+ */
+export class ArgSpec {
+  /**
+   * The canonical name of the argument. Corresponds to key in schema's `properties` prop.
+   * @type {string}
+   */
+  name;
+
+  /**
+   * The `ExtensionType` of the argument. This will be set if the arg came from an extension;
+   * otherwise it will be `undefined`.
+   * @type {ExtensionType|undefined}
+   */
+  extType;
+
+  /**
+   * The name of the extension, if this argument came from an extension.
+   *
+   * Otherwise `undefined`.
+   * @type {string|undefined}
+   */
+  extName;
+
+  /**
+   * The schema ID (`$id`) for the argument.  This is automatically determined, and any user-provided `$id`s will be overwritten.
+   *
+   * @type {string}
+   */
+  ref;
+
+  /**
+   * The CLI argument, sans leading dashes.
+   * @type {string}
+   */
+  arg;
+
+  /**
+   * The desired keypath for the argument after arguments have been parsed.
+   *
+   * Typically this is camelCased.  If the arg came from an extension, it will be prefixed with
+   * `<extType>.<extName>.`
+   * @type {string}
+   */
+  dest;
+
+  /**
+   * Whatever the default value of this argument is, as specified by the
+   * `default` property of the schema.
+   * @type {D}
+   */
+  defaultValue;
+
+  /**
+   * Builds some computed fields and assigns them to the instance.
+   *
+   * Undefined properties are not assigned.
+   *
+   * The _constructor_ is private. Use {@link ArgSpec.create} instead.
+   * @private
+   * @template D
+   * @param {string} name
+   * @param {ArgSpecOptions<D>} [opts]
+   */
+  constructor (name, {extType, extName, dest, defaultValue} = {}) {
+    // we must normalize the extension name to fit into our convention for CLI
+    // args.
+    const arg = ArgSpec.toArg(name, extType, extName);
+
+    const ref = ArgSpec.toSchemaRef(name, extType, extName);
+
+    // if no explicit `dest` provided, just camelCase the name to avoid needing
+    // to use bracket syntax when accessing props on the parsed args object.
+    const baseDest = _.camelCase(dest ?? name);
+
+    const destKeypath =
+      extType && extName ? [extType, extName, baseDest].join('.') : baseDest;
+
+    Object.defineProperties(this, {
```

Comment:
I think I probably did this because of TypeScript.  I maybe wanted to `return Object.freeze(this)` from the constructor, but it yelled at me.  I think, since we're using this factory method anyway, we can just return the frozen object from that.

Reply:
oh, I see what this does.  if one of the optional options is `undefined`, this will avoid enumerating the property; no `defaultValue: undefined`.  this probably does not matter either way, and I think I was just being fussy.

Reply:
I changed this to use `Object.freeze()`